bndtools.m2e/src/bndtools/m2e/MavenWorkspaceRepository.java

```diff
@@ -1,256 +1,375 @@
+					if (g.matcher(bsn)
+						.find()) {
+						bsns.add(bsn);
+						continue outer;
+					}
 				}
 			}
 		}
 
-		if (finalName == null) {
-			finalName = mavenProject.getBuild()
-				.getFinalName();
-		}
-
-		return new File(mavenProject.getBuild()
-			.getDirectory(), finalName + ".jar");
+		logger.debug("{}: list({}) {}", getName(), pattern, bsns);
+		return bsns;
 	}
 
-	private String getBsnFromMavenProject(MavenProject mavenProject) throws Exception {
-		final File bundleFile = getBundleFile(mavenProject);
-
-		if (bundleFile.exists()) {
-			Domain domain = Domain.domain(bundleFile);
-			String bsn = domain.getBundleSymbolicName()
-				.getKey();
-			return bsn;
+	@Override
+	public void mavenProjectChanged(MavenProjectChangedEvent[] events, IProgressMonitor monitor) {
+		for (MavenProjectChangedEvent event : ((events != null) ? events : new MavenProjectChangedEvent[0])) {
+			Kind kind = Kind.get(event);
+			logger.debug("{}: mavenProjectChanged({}, {})", getName(), kind, event.getSource());
+			switch (kind) {
+				case ADDED :
+				case CHANGED :
+					IMavenProjectFacade projectFacade = event.getMavenProject();
+					IProject iProject = projectFacade.getProject();
+					index(iProject, indexProject(projectFacade, monitor));
+					break;
+				case REMOVED :
+					remove(event.getOldMavenProject()
+						.getProject());
+					break;
+			}
 		}
 
-		return null;
+		list = Memoize.supplier(() -> list0());
+		unchecked(() -> Central.refreshPlugins());
 	}
 
-	private void init() {
-		inited = true;
-
-		final IProgressMonitor monitor = new NullProgressMonitor();
+	@Override
+	public PutResult put(InputStream stream, PutOptions options) throws Exception {
+		throw new IllegalStateException(getName() + " is read-only");
+	}
 
-		for (IMavenProjectFacade projectFacade : mavenProjectRegistry.getProjects()) {
-			try {
-				String bsnGuess = guessBsnFromProjectFacade(projectFacade);
+	@Override
+	public SortedSet<Version> versions(String bsn) throws Exception {
+		Map<Requirement, Collection<Capability>> providers = findProviders(bsnRequirements.apply(bsn));
+
+		SortedSet<Version> versions = providers.values()
+			.stream()
+			.flatMap(Collection<Capability>::stream)
+			.map(Capability::getAttributes)
+			.map(m -> m.get(IdentityNamespace.CAPABILITY_VERSION_ATTRIBUTE))
+			.map(String::valueOf)
+			.map(Version::new)
+			.collect(Collectors.toCollection(TreeSet<Version>::new));
+
+		logger.debug("{}: versions({}) {}", getName(), bsn, versions);
+		return versions;
+	}
 
-				if (bsnGuess != null) {
-					bsnMap.put(bsnGuess, projectFacade);
-				} else {
-					MavenProject mavenProject = getMavenProject(projectFacade, monitor);
-					String bsn = getBsnFromMavenProject(mavenProject);
+	@Override
+	protected BiFunction<ResourceBuilder, File, ResourceBuilder> fileIndexer(IProject project) {
+		String name = project.getName();
+		return (rb, file) -> {
+			// Add a capability specific to the workspace so that we can
```

Comment:
You need to all the super BiFunction or you need to call rb.addFile. You may want to just call rb.addFile yourself to get the hasIdentity value directly.

Reply:
fixed